File: Python/Medium/2657_find_the_prefix_common_array_of_two_arrays.py
"""
2657. Find the Prefix Common Array of Two Arrays

A and B are two 0-indexed permutations of 1..n. Return C where C[i] is the
count of integers present at or before index i in **both** A and B.

Problem Link: https://leetcode.com/problems/find-the-prefix-common-array-of-two-arrays/

Examples:
    Example 1:
        Input: A = [1, 3, 2, 4], B = [3, 1, 2, 4]
        Output: [0, 2, 3, 4]

    Example 2:
        Input: A = [2, 3, 1], B = [3, 1, 2]
        Output: [0, 1, 3]

Constraints:
    1 <= A.length == B.length == n <= 50
    1 <= A[i], B[i] <= n
    A and B are permutations of [1..n].
"""


class Solution:
    def findThePrefixCommonArray(self, A: list[int], B: list[int]) -> list[int]:
        """
        "Seen twice" frequency array. A value v becomes common to both
        prefixes the moment we see it for the second time (once in A, once
        in B — order doesn't matter, since each side is a permutation). One
        counter tracks how many such crossings have occurred.

        Recommended in Python — list-index updates beat the bitmask variant
        (see findThePrefixCommonArrayBitMask docstring for why).

        Time complexity:  O(n)
        Space complexity: O(n)
        """
        n: int = len(A)
        ans: list[int] = [0] * n
        seen: list[int] = [0] * (n + 1)
        counter: int = 0

        for i in range(n):
            seen[A[i]] += 1
            if seen[A[i]] == 2:
                counter += 1

            seen[B[i]] += 1
            if seen[B[i]] == 2:
                counter += 1

            ans[i] = counter

        return ans

    def findThePrefixCommonArrayBitMask(self, A: list[int], B: list[int]) -> list[int]:
        """
        Bitmask intersection. Bit v of maskA flips on once A has seen value
        v; same for maskB. (maskA & maskB).bit_count() is the size of the
        common prefix at step i.

        Algorithmically O(n), but in CPython it is typically *slower* than
        the frequency-array variant: Python ints are heap objects with
        arbitrary precision, so each `1 << A[i]` allocates a new integer
        and each `|=` reallocates the mask. The hardware-register speed of
        bitwise ops that makes this approach shine in Go/C++ is erased by
        the object overhead in Python.

        Time complexity:  O(n) — but with high per-op constants in CPython.
        Space complexity: O(n / word_size) for the masks.
        """
        n: int = len(A)
        ans: list[int] = [0] * n

        maskA = 0
        maskB = 0

        for i in range(n):
            maskA |= (1 << A[i])
            maskB |= (1 << B[i])

            ans[i] = (maskA & maskB).bit_count()

        return ans

    def findThePrefixCommonArraySet(self, A: list[int], B: list[int]) -> list[int]:
        """
        Maintain two growing sets and recompute the intersection size each
        step. Cleanest to read, but the per-step intersection is O(i), so
        total cost is O(n^2). Useful as a baseline / sanity-check.

        Time complexity:  O(n^2)
        Space complexity: O(n)
        """
        n: int = len(A)
        ans: list[int] = [0] * n

        setA: set[int] = set()
        setB: set[int] = set()

        for i in range(n):
            setA.add(A[i])
            setB.add(B[i])
            ans[i] = len(setA.intersection(setB))

        return ans


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ([1, 3, 2, 4], [3, 1, 2, 4], [0, 2, 3, 4]),
        ([2, 3, 1], [3, 1, 2], [0, 1, 3]),
        ([1], [1], [1]),
        ([1, 2], [2, 1], [0, 2]),
    ]

    for i, (A, B, expected) in enumerate(test_cases, 1):
        assert solution.findThePrefixCommonArray(list(A), list(B)) == expected, f"freq test {i} failed"
        assert solution.findThePrefixCommonArrayBitMask(list(A), list(B)) == expected, f"bitmask test {i} failed"
        assert solution.findThePrefixCommonArraySet(list(A), list(B)) == expected, f"set test {i} failed"

    print("All test cases passed!")

File: Golang/Medium/2657_find_the_prefix_common_array_of_two_arrays.go
package main

import (
	"fmt"
	"math/bits"
)

// findThePrefixCommonArray returns C where C[i] counts the integers present
// in both A[0..i] and B[0..i]. A and B are permutations of 1..n with n <= 50.
//
// Bitmask intersection. maskA / maskB each carry a bit per value seen so
// far on their side; (maskA & maskB) is the running intersection, and
// bits.OnesCount64 reads its size in one instruction. n <= 50 fits in a
// single uint64, so the whole prefix lives in two CPU registers.
//
// Time complexity:  O(n)
// Space complexity: O(1) extra (besides the output slice).
func findThePrefixCommonArray(A []int, B []int) []int {
	n := len(A)
	ans := make([]int, n)

	var maskA, maskB uint64

	for i := 0; i < n; i++ {
		maskA |= (1 << A[i])
		maskB |= (1 << B[i])

		ans[i] = bits.OnesCount64(maskA & maskB)
	}
	return ans
}

// findThePrefixCommonArraySeenTwice is the frequency-array variant: any
// value seen twice (once in A, once in B — permutation guarantees no
// duplicates within a side) is common. A single counter tracks how many
// such crossings have occurred.
//
// Time complexity:  O(n)
// Space complexity: O(n)
func findThePrefixCommonArraySeenTwice(A []int, B []int) []int {
	n := len(A)
	ans := make([]int, n)

	seen := make([]int, n+1)
	commonCount := 0

	for i := 0; i < n; i++ {
		seen[A[i]]++
		if seen[A[i]] == 2 {
			commonCount++
		}

		seen[B[i]]++
		if seen[B[i]] == 2 {
			commonCount++
		}

		ans[i] = commonCount
	}

	return ans
}

// findThePrefixCommonArraySet is the map-set baseline — clearest to read,
// quadratic to run because the per-step intersection scans the smaller set.
//
// Time complexity:  O(n^2)
// Space complexity: O(n)
func findThePrefixCommonArraySet(A []int, B []int) []int {
	n := len(A)
	ans := make([]int, n)
	setA := make(map[int]struct{})
	setB := make(map[int]struct{})

	for i := 0; i < n; i++ {
		setA[A[i]] = struct{}{}
		setB[B[i]] = struct{}{}
		ans[i] = len(intersect(setA, setB))
	}

	return ans
}

func intersect(set1, set2 map[int]struct{}) map[int]struct{} {
	intersection := make(map[int]struct{})

	if len(set1) > len(set2) {
		set1, set2 = set2, set1
	}

	for item := range set1 {
		if _, exists := set2[item]; exists {
			intersection[item] = struct{}{}
		}
	}
	return intersection
}

func main() {
	testCases := []struct {
		A        []int
		B        []int
		expected []int
	}{
		{[]int{1, 3, 2, 4}, []int{3, 1, 2, 4}, []int{0, 2, 3, 4}},
		{[]int{2, 3, 1}, []int{3, 1, 2}, []int{0, 1, 3}},
		{[]int{1}, []int{1}, []int{1}},
		{[]int{1, 2}, []int{2, 1}, []int{0, 2}},
	}

	funcs := []struct {
		label string
		fn    func([]int, []int) []int
	}{
		{"findThePrefixCommonArray           ", findThePrefixCommonArray},
		{"findThePrefixCommonArraySeenTwice  ", findThePrefixCommonArraySeenTwice},
		{"findThePrefixCommonArraySet        ", findThePrefixCommonArraySet},
	}

	for _, f := range funcs {
		for i, tc := range testCases {
			result := f.fn(tc.A, tc.B)
			status := "PASS"
			if fmt.Sprintf("%v", result) != fmt.Sprintf("%v", tc.expected) {
				status = "FAIL"
			}
			fmt.Printf("%s | Example %d | A=%v B=%v | Expected: %v | Result: %v | [%s]\n",
				f.label, i+1, tc.A, tc.B, tc.expected, result, status)
		}
	}
}
